apps/wing/src/cli.ts

```diff
@@ -65,6 +67,13 @@ async function main() {
         .default("sim")
     )
     .option("-p, --plugins [plugin...]", "Compiler plugins")
+    .hook("preAction", async (cmd) => {
+      const target = cmd.opts().target;
+      const progress = program.opts().progress;
+      if (progress !== false && target !== "sim") {
+        process.env.PROGRESS = "1";
+      }
+    })
```

Comment:
Okay so it is possible to add this to the top level program, I had to do a little digging though because commander's API is kinda confusing:

```ts
program.hook("preSubcommand", (_: Command, cmd: Command) => {
	const options = cmd.optsWithGlobals();
  const target = options.target;
  const progress = options.progress;
  if (progress !== false && target !== "sim") {
    process.env.PROGRESS = "1";
  }
});
```

(note the `preSubcommand`)

With this you'll be able to have this only once, with access to both target and progress.

Reply:
There is a problem with your solution I think. The "preSubcommand" `HookEvent` gets executed before the `test` and `compile` commands. For some reasons the value of `target` option which you specify in the terminal doesn't get passed to the "Subcommand", resulting in an `undefined` target value, which ends up setting the default value of `sim`.

```ts
program.hook("preSubcommand", (_: Command, cmd: Command) => {
  const options = cmd.optsWithGlobals();
  const target = options.target;
  const progress = options.progress;
  log(options);
  log(target);
  if (!progress && target !== "sim") {
    process.env.PROGRESS = "1";
  }
});
```

```console
$ mywing test -t tf-aws examples/tests/sdk_tests/function/invoke.w
```

```console
{ target: 'sim', updateCheck: true }
sim
```

Reply:
There is a problem with your solution I think. The "preSubcommand" `HookEvent` gets executed before the `test` and `compile` commands. For some reasons, the value of `target` option which you specify in the terminal doesn't get passed to the "Subcommand", resulting in an `undefined` target value, which ends up setting the default value of `sim`.

```ts
program.hook("preSubcommand", (_: Command, cmd: Command) => {
  const options = cmd.optsWithGlobals();
  const target = options.target;
  const progress = options.progress;
  log(options);
  log(target);
  if (!progress && target !== "sim") {
    process.env.PROGRESS = "1";
  }
});
```

```console
$ mywing test -t tf-aws examples/tests/sdk_tests/function/invoke.w
```

```console
{ target: 'sim', updateCheck: true }
sim
```

Reply:
There is a problem with your solution I think. The "preSubcommand" `HookEvent` gets executed before the `test` and `compile` commands. For some reason the value of `target` option which you specify in the terminal doesn't get passed to the "Subcommand", resulting in an `undefined` target value, which ends up setting the default value of `sim`.

```ts
program.hook("preSubcommand", (_: Command, cmd: Command) => {
  const options = cmd.optsWithGlobals();
  const target = options.target;
  const progress = options.progress;
  log(options);
  log(target);
  if (!progress && target !== "sim") {
    process.env.PROGRESS = "1";
  }
});
```

```console
$ mywing test -t tf-aws examples/tests/sdk_tests/function/invoke.w
```

```console
{ target: 'sim', updateCheck: true }
sim
```

Reply:
Ah okay, fair enough. Sorry for hitting on this so hard and thinks for figuring out the root issue, just really wanted to avoid duplication if possible.

I still think my fallback option is possible, but I'll approve the change regardless. Basically, it is still possible to extract a function:

```ts
const program = new Command();

// ...

const progressHook = (cmd: Command) => {
  const target = cmd.opts().target;
  const progress = program.opts().progress;
  if (progress !== false && target !== "sim") {
    process.env.PROGRESS = "1";
  }
}
```

Then you can use the same function in both hooks (`program` is captured in the scope):

```ts
program.command("compile")

// ...

.hook("preAction", progressHook)
```

Reply:
Thanks for the suggestion, I honestly didn't think about. I removed the duplicated code.

Reply:
Thanks for the suggestion, I honestly didn't think about that. I removed the duplicated code.